File: backend/services/financial_calculator.py
from dataclasses import dataclass
from typing import Dict, Any, Optional
import math

class FinancialCalculator:
    """
    Utility class for performing complex financial calculations.
    """

    @staticmethod
    def calculate_savings_rate(income: float, expenses: float) -> float:
        """
        Calculate savings rate percentage.
        Formula: ((Income - Expenses) / Income) * 100
        """
        if income <= 0:
            return 0.0
        return ((income - expenses) / income) * 100

    @staticmethod
    def calculate_runway(current_savings: float, monthly_expenses: float) -> float:
        """
        Calculate financial runway in months.
        Formula: Current Savings / Monthly Expenses
        """
        if monthly_expenses <= 0:
            return float('inf')  # Infinite runway if no expenses
        return current_savings / monthly_expenses

    @staticmethod
    def calculate_compound_interest(
        principal: float, 
        annual_rate: float, 
        years: int, 
        monthly_contribution: float = 0
    ) -> Dict[str, float]:
        """
        Calculate compound interest including monthly contributions.
        Returns: { 'total_amount': float, 'interest_earned': float, 'total_contributed': float }
        """
        r = annual_rate / 100
        n = 12  # Compounded monthly
        t = years
        
        # Future value of principal
        future_value_principal = principal * (1 + r/n)**(n*t)
        
        # Future value of monthly contributions
        # FV = PMT * (((1 + r/n)^(n*t) - 1) / (r/n))
        if r > 0:
            future_value_contributions = monthly_contribution * (((1 + r/n)**(n*t) - 1) / (r/n))
        else:
            future_value_contributions = monthly_contribution * n * t
            
        total_amount = future_value_principal + future_value_contributions
        total_contributed = principal + (monthly_contribution * n * t)
        interest_earned = total_amount - total_contributed
        
        return {
            "total_amount": round(total_amount, 2),
            "interest_earned": round(interest_earned, 2),
            "total_contributed": round(total_contributed, 2)
        }

    @staticmethod
    def calculate_per_capita_income(total_income: float, family_size: int) -> float:
        """
        Calculate Per Capita Income (Income per family member).
        """
        if family_size <= 0:
            return 0.0
        return total_income / family_size

    @staticmethod
    def calculate_loan_emi(principal: float, annual_rate: float, tenure_years: float) -> float:
        """
        Calculate Monthly EMI for a loan.
        Formula: E = P * r * (1+r)^n / ((1+r)^n - 1)
        """
        if principal <= 0 or tenure_years <= 0:
            return 0.0
        if annual_rate <= 0:
            return principal / (tenure_years * 12)
            
        r = annual_rate / (12 * 100)  # Monthly rate
        n = tenure_years * 12         # Total months
        
        emi = principal * r * ((1 + r)**n) / (((1 + r)**n) - 1)
        return round(emi, 2)

    @staticmethod
    def calculate_fire_number(annual_expenses: float, withdrawal_rate: float = 4.0) -> float:
        """
        Calculate FIRE (Financial Independence, Retire Early) number.
        Target corpus needed to sustain annual expenses indefinitely.
        Formula: Annual Expenses * (100 / Withdrawal Rate) (Default 25x rule)
        """
        if withdrawal_rate <= 0:
            return float('inf')
        return annual_expenses * (100 / withdrawal_rate)

    @staticmethod
    def calculate_emergency_fund_status(
        current_savings: float, 
        monthly_expenses: float, 
        target_months: int = 6
    ) -> Dict[str, Any]:
        """
        Evaluate Emergency Fund status.
        """
        target_amount = monthly_expenses * target_months
        if target_amount == 0:
             return {"status": "Complete", "percentage": 100, "shortfall": 0}
             
        percent_complete = (current_savings / target_amount) * 100
        shortfall = max(0, target_amount - current_savings)
        
        status = "Critical"
        if percent_complete >= 100:
            status = "Excellent"
        elif percent_complete >= 80:
            status = "Good"
        elif percent_complete >= 50:
            status = "Fair"
        elif percent_complete >= 20:
            status = "Poor"
            
        return {
            "target_amount": target_amount,
            "current_amount": current_savings,
            "shortfall": shortfall,
            "percentage": round(percent_complete, 1),
            "health_status": status,
            "months_covered": round(current_savings / monthly_expenses, 1) if monthly_expenses > 0 else 0
        }

File: frontend/wealthin_flutter/lib/core/services/financial_calculator.dart
import 'dart:math' as math;

/// Financial Calculator Service
/// Pure Dart implementation of financial formulas.
/// Replaces the Python backend calculator service for instant, offline results.
class FinancialCalculator {
  
  // ==================== INVESTMENT CALCULATORS ====================

  /// Calculate SIP (Systematic Investment Plan) Returns
  static Map<String, dynamic> calculateSIP({
    required double monthlyInvestment,
    required double expectedRate, // Annual rate in %
    required int durationMonths,
  }) {
    final monthlyRate = expectedRate / 12 / 100;
    
    double futureValue;
    if (monthlyRate == 0) {
      futureValue = monthlyInvestment * durationMonths;
    } else {
      // FV = P × ((1 + r)^n - 1) / r × (1 + r)
      futureValue = monthlyInvestment * 
                   ((math.pow(1 + monthlyRate, durationMonths) - 1) / monthlyRate) * 
                   (1 + monthlyRate);
    }

    final totalInvested = monthlyInvestment * durationMonths;
    final wealthGained = futureValue - totalInvested;

    return {
      'monthly_investment': monthlyInvestment,
      'duration_months': durationMonths,
      'expected_rate': expectedRate,
      'total_invested': totalInvested.roundToDouble(),
      'future_value': futureValue.roundToDouble(),
      'wealth_gained': wealthGained.roundToDouble(),
      'returns_percentage': totalInvested > 0 
          ? ((wealthGained / totalInvested) * 100).toStringAsFixed(2) 
          : "0.00"
    };
  }

  /// Calculate FD (Fixed Deposit) Maturity
  static Map<String, dynamic> calculateFD({
    required double principal,
    required double rate, // Annual rate in %
    required int tenureMonths,
    String compounding = "quarterly", // Not used in simple formula, assuming quarterly by default standards
  }) {
    // Standard quarterly compounding for Indian FDs
    const n = 4.0; 
    final r = rate / 100;
    final t = tenureMonths / 12;

    // A = P(1 + r/n)^(nt)
    final maturityAmount = principal * math.pow(1 + r / n, n * t);
    final interestEarned = maturityAmount - principal;
    final effectiveAnnualRate = (math.pow(1 + r / n, n) - 1) * 100;

    return {
      'principal': principal,
      'rate': rate,
      'tenure_months': tenureMonths,
      'maturity_amount': maturityAmount.roundToDouble(),
      'interest_earned': interestEarned.roundToDouble(),
      'effective_annual_rate': effectiveAnnualRate.toStringAsFixed(2)
    };
  }

  /// Calculate RD (Recurring Deposit) Maturity
  static Map<String, dynamic> calculateRD({
    required double monthlyDeposit,
    required double rate, // Annual rate in %
    required int tenureMonths,
  }) {
    // RD Formula with Quarterly Compounding
    final quarterlyRate = rate / 4 / 100;
    double maturityAmount = 0;

    for (int month = 0; month < tenureMonths; month++) {
      // Each installment earns interest for the remaining period
      // This is a simplified iterative approach accurate for standard RDs
      final remainingQuarters = (tenureMonths - month) / 3;
      final amount = monthlyDeposit * math.pow(1 + quarterlyRate, remainingQuarters);
      maturityAmount += amount;
    }

    final totalDeposited = monthlyDeposit * tenureMonths;
    final interestEarned = maturityAmount - totalDeposited;

    return {
      'monthly_deposit': monthlyDeposit,
      'rate': rate,
      'tenure_months': tenureMonths,
      'maturity_amount': maturityAmount.roundToDouble(),
      'total_deposited': totalDeposited.roundToDouble(),
      'interest_earned': interestEarned.roundToDouble()
    };
  }

  /// Calculate Compound Interest (Lumpsum with optional monthly contribution)
  static Map<String, dynamic> calculateCompoundInterest({
    required double principal,
    required double rate, // Annual rate in %
    required int years,
    double monthlyContribution = 0,
  }) {
    final r = rate / 100;
    const n = 12; // Monthly compounding standard
    final t = years;

    // Future Value of Initial Principal
    final fvPrincipal = principal * math.pow(1 + r / n, n * t);

    // Future Value of Monthly Contributions
    // FV = PMT * (((1 + r/n)^(n*t) - 1) / (r/n))
    double fvContributions = 0;
    if (monthlyContribution > 0) {
      if (r > 0) {
        fvContributions = monthlyContribution * (((math.pow(1 + r / n, n * t) - 1) / (r / n)));
      } else {
        fvContributions = monthlyContribution * n * t;
      }
    }

    final totalAmount = fvPrincipal + fvContributions;
    final totalContributed = principal + (monthlyContribution * n * t);
    final interestEarned = totalAmount - totalContributed;

    return {
      'total_amount': totalAmount.roundToDouble(),
      'interest_earned': interestEarned.roundToDouble(),
      'total_contributed': totalContributed.roundToDouble(),
      'years': years
    };
  }

  // ==================== LOAN CALCULATORS ====================

  /// Calculate EMI (Equated Monthly Installment)
  static Map<String, dynamic> calculateEMI({
    required double principal,
    required double rate, // Annual rate in %
    required int tenureMonths,
  }) {
    final monthlyRate = rate / 12 / 100;
    
    double emi;
    if (monthlyRate == 0) {
      emi = principal / tenureMonths;
    } else {
      // E = P * r * (1+r)^n / ((1+r)^n - 1)
      emi = principal * monthlyRate * 
            (math.pow(1 + monthlyRate, tenureMonths)) / 
            ((math.pow(1 + monthlyRate, tenureMonths) - 1));
    }

    final totalPayment = emi * tenureMonths;
    final totalInterest = totalPayment - principal;

    return {
      'principal': principal,
      'rate': rate,
      'tenure_months': tenureMonths,
      'emi': emi.roundToDouble(),
      'total_payment': totalPayment.roundToDouble(),
      'total_interest': totalInterest.roundToDouble()
    };
  }

  // ==================== FINANCIAL HEALTH CALCULATORS ====================

  /// Calculate Savings Rate (%)
  static double calculateSavingsRate(double income, double expenses) {
    if (income <= 0) return 0.0;
    final rate = ((income - expenses) / income) * 100;
    return double.parse(rate.toStringAsFixed(2));
  }

  /// Calculate Per Capita Income (Income per family member)
  static double calculatePerCapitaIncome(double totalIncome, int familySize) {
    if (familySize <= 0) return 0.0;
    return double.parse((totalIncome / familySize).toStringAsFixed(2));
  }

  /// Evaluate Emergency Fund Status
  static Map<String, dynamic> calculateEmergencyFundStatus({
    required double currentSavings,
    required double monthlyExpenses,
    int targetMonths = 6,
  }) {
    final targetAmount = monthlyExpenses * targetMonths;
    if (targetAmount == 0) {
      return {
        "status": "Complete", 
        "percentage": 100.0, 
        "shortfall": 0.0,
        "months_covered": 0.0,
        "health_status": "Complete"
      };
    }

    final percentComplete = (currentSavings / targetAmount) * 100;
    final shortfall = math.max(0.0, targetAmount - currentSavings);
    final monthsCovered = monthlyExpenses > 0 ? (currentSavings / monthlyExpenses) : 0.0;

    String status = "Critical";
    if (percentComplete >= 100) {
      status = "Excellent";
    } else if (percentComplete >= 80) status = "Good";
    else if (percentComplete >= 50) status = "Fair";
    else if (percentComplete >= 20) status = "Poor";

    return {
      "target_amount": targetAmount,
      "current_amount": currentSavings,
      "shortfall": shortfall,
      "percentage": double.parse(percentComplete.toStringAsFixed(1)),
      "health_status": status,
      "months_covered": double.parse(monthsCovered.toStringAsFixed(1))
    };
  }
}
